### PR TITLE
[isoltest] Fix (aligned) hex parsing and formatting

### LINKED ISSUE #6298: [soltest] Format expectations as compact hex literals by default.
When reformatting a hex literal like ``0x42``, isoltest will currently correctly output a hex literal, but always a full 32-bytes long one like (at least confirmed on mismatching expectations), e.g. ``0x000000000000000000000000000000000000000000000000000000000000002a``. This is "fine", but it would be nice to just change the formatting to use "toCompactHexWithPrefix", which will result in a nice ``0x2a``.

On a related note (but probably not worth its own issue): parsing of hex literals should be more relaxed: currently odd numbers of digits are considered invalid literals, e.g. ``0x01`` is fine, but ``0x1`` and ``0x101`` are invalid and instead you have to use ``0x01`` or ``0x0101``. This is inherited from solidity, where it makes sense, but here I think it is more cumbersome than useful.

### LINKED ISSUE #7082: [semantic test] Report concrete reason of invalid encoding to user
When I was adding new semantic tests, I tried to add a test case that contains `0x140` (which is an illegal literal because hex number needs to have even length, for some reason), and the test framework just says `Literal encoding invalid.`, which is not very helpful and confuses me.

It could be useful if it prints more concrete reasons. Also seems like `TestFileParser::parseABITypeLiteral()` is not properly passing (rethrow) error reason to caller?

I can try to fix this when I get time.

### PR DESCRIPTION
**Pull request by erak**
Depends on https://github.com/ethereum/solidity/pull/7052.
- [x] Change target branch after the PR above got merged

Fixes https://github.com/ethereum/solidity/issues/6298.
Fixes https://github.com/ethereum/solidity/issues/7082.

### CODE REVIEW

File: test/libsolidity/util/TestFileParser.cpp [hunk 1]
```diff
@@ -226,104 +226,97 @@ Parameter TestFileParser::parseParameter()
...
+			throw Error(Error::Type::ParserError, "Hex string literals cannot be aligned or padded.");
+
+		string parsed = parseString();
+		parameter.rawString += "hex\"" + parsed + "\"";
+		parameter.rawBytes = BytesUtils::convertHexNumber(parsed);
+		parameter.abiType = ABIType{
+			ABIType::HexString, ABIType::AlignNone, parameter.rawBytes.size()
+		};
 	}
-	catch (std::exception const&)
```
Comment (leonardoalt) [inline test/libsolidity/util/TestFileParser.cpp:350]: Where is it handled now?

Comment (erak) [reply test/libsolidity/util/TestFileParser.cpp:350]: It was always handled in isoltest: https://github.com/ethereum/solidity/blob/develop/test/tools/isoltest.cpp#L185. Here, at this point it was just not correct to catch exceptions already and swallow the root cause by creating a new generic exception.

**Comment by leonardoalt** [comment]
Please rebase

**Comment by erak** [comment]
@sifmelcara, you may want to have a look since you've reported one of the issues this PR fixes ;)

File: test/libsolidity/util/BytesUtils.cpp [hunk 2]
```diff
@@ -159,13 +158,8 @@ string BytesUtils::formatBoolean(bytes const& _bytes)
...
 string BytesUtils::formatHex(bytes const& _bytes)
 {
-	stringstream os;
-
-	string hex{toHex(_bytes, HexPrefix::Add)};
-	boost::algorithm::replace_all(hex, "00", "");
-	os << hex;
-
-	return os.str();
+	u256 value = fromBigEndian<u256>(_bytes);
```
Comment (bshastry) [inline test/libsolidity/util/BytesUtils.cpp:162]: Why is `_bytes` being cropped to `u256.` Is it the case that input to `formatHex` is at most 32 bytes?

File: test/libsolidity/util/TestFunctionCallTests.cpp [hunk 3]
```diff
@@ -139,7 +139,8 @@ BOOST_AUTO_TEST_CASE(format_hex_singleline)
 	BOOST_REQUIRE_EQUAL(test.format(), "// f(bytes32): 0x31 -> 0x31");
 
 	bytes actualResult = fromHex("0x32");
-	bytes actualBytes = actualResult + bytes(32 - actualResult.size(), 0);
+	bytes actualBytes = bytes(32 - actualResult.size(), 0) + actualResult;
```
Comment (bshastry) [inline test/libsolidity/util/TestFunctionCallTests.cpp:142]:
```suggestion
	bytes actualBytes += bytes(32 - actualResult.size(), 0);
```

File: test/libsolidity/util/TestFileParserTests.cpp [hunk 4]
```diff
@@ -786,14 +786,6 @@ BOOST_AUTO_TEST_CASE(call_ether_type_invalid)
...
 }
 
-BOOST_AUTO_TEST_CASE(call_hex_number_invalid)
-{
-	char const* source = R"(
-		// f(bytes32, bytes32): 0x616, 0x042 -> 1
-	)";
-	BOOST_REQUIRE_THROW(parse(source), langutil::Error);
-}
-
```
Comment (bshastry) [inline test/libsolidity/util/TestFileParserTests.cpp:796]: If I understood correctly, can this be changed to a `BOOST_REQUIRE_SUCCESS`?

File: test/libsolidity/util/TestFileParser.cpp [hunk 5]
```diff
@@ -252,104 +252,97 @@ Parameter TestFileParser::parseParameter()
...
+	else if (accept(Token::Hex, true))
+	{
+		if (isSigned)
+			throw Error(Error::Type::ParserError, "Invalid hex string literal.");
 		if (parameter.alignment != Parameter::Alignment::None)
-		{
-			expect(Token::RParen);
-			parameter.rawString += formatToken(Token::RParen);
-		}
+			throw Error(Error::Type::ParserError, "Hex string literals cannot be aligned or padded.");
```
Comment (bshastry) [inline test/libsolidity/util/TestFileParser.cpp:293]: iiuc, apart from removing the `try...catch` block, this is the most important change, right? That hex literals can no longer be padded/aligned?

Comment (bshastry) [reply test/libsolidity/util/TestFileParser.cpp:293]: Does it make sense to add a test case that triggers the newly added `throw`?

**Review by bshastry — COMMENTED** [review]
LGTM apart from a couple of minor comments

File: test/libsolidity/util/BytesUtils.cpp (hunk 2 above)
Comment (chriseth) [reply test/libsolidity/util/BytesUtils.cpp:162]: Please at least add a check about the length of `_bytes`. What happens if it is shorter than 32 bytes and starts with zeros?

Comment (erak) [reply test/libsolidity/util/BytesUtils.cpp:162]: @bshastry Yes, it's supposed to be called with at most 32 bytes. I added an assertion for that.
@chriseth It's being cropped: e.g. `bytes{0,0,0,3,255,0,2}` becomes `0x3FF0002`

File: test/libsolidity/util/TestFunctionCallTests.cpp (hunk 3 above)
Comment (erak) [reply test/libsolidity/util/TestFunctionCallTests.cpp:142]: Mmh, I still need the `actualResult`, which is different from `actualBytes`.

File: test/libsolidity/util/TestFileParserTests.cpp (hunk 4 above)
Comment (erak) [reply test/libsolidity/util/TestFileParserTests.cpp:796]: Ah, nice one. I added another test based on the one removed which checks oddly-sized hex number literals.

File: test/libsolidity/util/BytesUtils.cpp (hunk 2 above)
Comment (chriseth) [reply test/libsolidity/util/BytesUtils.cpp:162]: @erak if that's what we want then it's fine.

File: test/libsolidity/util/TestFileParser.cpp (hunk 5 above)
Comment (erak) [reply test/libsolidity/util/TestFileParser.cpp:293]: Unfortunately, the diff looks a bit messed up. I just removed `try...catch` block and this `throw` was already there: https://github.com/ethereum/solidity/pull/7091/files#diff-9a9ec295472c0ac89c7f1c1f0cabf58cL295.

The change is that exceptions thrown here are not caught immediately any more (which was a defect), but rather handled by isoltest (which now outputs the cause for root exception).

File: test/libsolidity/util/TestFileParserTests.cpp [hunk 6]
```diff
@@ -404,6 +404,27 @@ BOOST_AUTO_TEST_CASE(call_arguments_string)
 	);
 }
 
+
+BOOST_AUTO_TEST_CASE(call_hex_number)
```
Comment (erak) [inline test/libsolidity/util/TestFileParserTests.cpp:408]: @bshastry This is the new test.

File: test/libsolidity/util/BytesUtils.cpp (hunk 2 above)
Comment (erak) [reply test/libsolidity/util/BytesUtils.cpp:162]: I think so. But I'm also fine with asserting that this function is only called with an input of 32 bytes. Unpadded values should be handled as hex strings in any case.

**Review by chriseth — APPROVED** [review]

**Pull request merged**